com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java

```diff
@@ -136,6 +146,109 @@ private int buildEmbedded(IResource input, CollectionDesc.Builder builder, Task<
+            ProtoUtil.merge(res, builder);
+
+            countComponentsInCollectionRecursively(project, builder, componentsInFactories);
+        }
+    }
+
+    private boolean isFactoryType(String type) throws IOException, CompileExceptionError {
+        return type.equals("factory") || type.equals("collectionfactory");
+    }
+
+    private void countComponentInResource(Project project, IResource res, Map<String, Integer> target) throws IOException, CompileExceptionError {
+        PrototypeDesc.Builder prot = PrototypeDesc.newBuilder();
+        ProtoUtil.merge(res, prot);
+
+        for (EmbeddedComponentDesc cd : prot.getEmbeddedComponentsList()) {
+            String type = cd.getType();
+            addOneComponent(type, target);
+            if (isFactoryType(type)) {
+                byte[] data = cd.getData().getBytes();
+                long hash = MurmurHash.hash64(data, data.length);
+                IResource resource = project.getGeneratedResource(hash, type);
+                resource.setContent(data);
+                countComponentsInFactory(project, resource, type);
+            }
+        }
+        for (ComponentDesc cd : prot.getComponentsList()) {
+            String comp = cd.getComponent();
+            String type = FilenameUtils.getExtension(comp);
+            addOneComponent(type, target);
+            if (isFactoryType(type)) {
+                IResource resource = project.getResource(comp);
+                countComponentsInFactory(project, resource, type);
+            }
+        }
+    }
+
+    private void countComponentsInCollectionRecursively(Project project, CollectionDesc.Builder builder, Map<String, Integer> target) throws IOException, CompileExceptionError {
+        for (InstanceDesc inst : builder.getInstancesList()) {
+            IResource res = project.getResource(inst.getPrototype());
+            countComponentInResource(project, res, target);
+        }
+        for (EmbeddedInstanceDesc desc : builder.getEmbeddedInstancesList()) {
+            byte[] data = desc.getData().getBytes();
+            long hash = MurmurHash.hash64(data, data.length);
+
+            IResource res = project.getGeneratedResource(hash, "go");
+            countComponentInResource(project, res, target);
+        }
+        for (CollectionInstanceDesc collInst : builder.getCollectionInstancesList()) {
+            IResource collResource = project.getResource(collInst.getCollection());
+            CollectionDesc.Builder subCollBuilder = CollectionDesc.newBuilder();
+            ProtoUtil.merge(collResource, subCollBuilder);
+            countComponentsInCollectionRecursively(project, subCollBuilder, target);
+        }
+    }
+
+    private String replaceComponentName(Project project, String in) {
+        return project.replaceExt("." + in).substring(1);
+    }
+
+    private void countAllComponentTypes(Project project, CollectionDesc.Builder builder) throws IOException, CompileExceptionError {
```

Comment:
Naming is much better and readable now. Thank you!

Reply:
No, Thank you! )